### PR TITLE
issue-109: timeline on * index pattern AND issue-33: ES query should retrieve the last X (limit) documents

### LINKED ISSUE #33: ES query should retrieve the last X (limit) documents
Hi,

In my opinion the ES query generated by your plugin should retrieve the **last** X (limit - default 100) documents matching the criteria and not the **first** X documents as it's the case actually because in my opinion the current view could easily be misleading while using the timeline to track/display the latest events.

Or even better, we could have an option in the "Option" tab to select if we want our limit to be ascending or descending. 

Thanks.

### LINKED ISSUE #109: timeline on * index pattern
1. Create an index pattern *
2. Create a saved search on index pattern *
3. Create a timeline visualization
4. Select one of the date fields available
5. Run the visualization (green button)

image

### PR DESCRIPTION
**Pull request by giorgio94c**
Close #109 
Close #33 

Probably there is a better way to write this but, at least, now it works!

### CODE REVIEW

**Comment by szydan** [comment]
@giorgio94c could you describe here a bit how sorting helps in case of * pattern, please.

**Comment by giorgio94c** [comment]
Sure @szydan : the problem with a saved search based on _* query_ is that we have all possible values in a specific field, from something to `undefined`. With sort I want to order all this values in the specific field we want to use as _start_ or _end_ date to prevent possibility of `undefined` records when we extract the first _n values_ (where _n_ means the value in _Events number limit_, default 100)

**Comment by szydan** [comment]
@giorgio94c I think it would be good to add such explanation to the place in the code where you do this sorting so the next person will know the reason behind this 

Now I remember that there was one person asking for such a feature to be able to sort here
https://github.com/sirensolutions/kibi_timeline_vis/issues/33

Could we make it an option which by default is turned on 
and give the user ability to change the sorting as in some cases the default sorting might not be what user wants

**Comment by giorgio94c** [comment]
> Could we make it an option which by default is turned on

Ok @szydan but In **visualization** or **Kibi > Advanced Settings**? And if the first case, for every _group_ (1 checkbox + 1 menu with _"asc"_ or _"desc"_) or _all groups_ (in _"Advanced"_)?

**Comment by szydan** [comment]
@giorgio94c I think in visualisation as it can be different per group
I would add 1 select with options like ( none, startField asc, startField desc, endField asc, endFiled desc

**Comment by giorgio94c** [comment]
> I would add 1 select with options like ( none, startField asc, startField desc, endField asc, endFiled desc

Ok, what's the default value? And I work on the other one issue (so we close this one) you linked before?

**Comment by szydan** [comment]
@giorgio94c  I would say let's try by default to show the last N recent events first

**Comment by giorgio94c** [comment]
Ok @szydan ! Just two things: I do this changes still in this issue? And with _None_ you mean a _blanck option_ in menu or you want a _"None" choice_?

File: public/lib/helpers/timeline_helper.js [hunk 1]
```diff
@@ -122,6 +122,22 @@ define(function (require) {
       return sortObj;
     };
 
+    /**
+     * Creates an Elasticsearch sort object to sort in chronological order on end field
+     *
+     * @param params group configuraton parameters
+     * @returns Elasticsearch sort object
+     */
+    TimelineHelper.prototype.getSortOnEndFieldObject = function (params) {
```
Comment (SergeyBondarenko) [inline public/lib/helpers/timeline_helper.js]: This method is completely equal to the `getSortOnStartFieldObject` method above. Don't duplicate the code, use one method instead.

**Comment by szydan** [comment]
I think it would be the best to do the two in this single PR 
Yes None is like blank option

File: public/lib/helpers/timeline_helper.js [hunk 2]
```diff
@@ -122,6 +122,22 @@ define(function (require) {
...
+      const sortObj = {};
+      if (params.endFieldSequence) {
+        sortObj[params.endFieldSequence.join('.')] = { order: 'asc' };
```
Comment (SergeyBondarenko) [inline public/lib/helpers/timeline_helper.js]: It is better to pass the order (`asc`) as a parameter to this method. In this way, you make it easier to read the code in the https://github.com/sirensolutions/kibi_timeline_vis/blob/master/public/kibi_timeline.js

File: public/lib/helpers/timeline_helper.js (hunk 1 above)
Comment (giorgio94c) [inline public/lib/helpers/timeline_helper.js]: All these things are going to be changed due to discussion with @szydan in this PR ;)

File: public/lib/helpers/timeline_helper.js (hunk 2 above)
Comment (giorgio94c) [inline public/lib/helpers/timeline_helper.js]: All these things are going to be changed due to discussion with @szydan in this PR ;)

**Comment by giorgio94c** [comment]
jenkins test it

**Comment by giorgio94c** [comment]
jenkins test it

**Comment by szydan** [comment]
jenkins test it

**Comment by giorgio94c** [comment]
jenkins test it

**Comment by giorgio94c** [comment]
Problem with tests because now, with latest update, we need to import `moment-timezone` instead of `moment`...
Now should be ok!

**Pull request merged**